gpu/test/matrix/ell_kernels.cpp

```diff
@@ -0,0 +1,165 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <core/matrix/ell.hpp>
+
+
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <core/base/exception.hpp>
+#include <core/base/exception_helpers.hpp>
+#include <core/base/executor.hpp>
+#include <core/matrix/dense.hpp>
+#include <core/test/utils.hpp>
+
+
+namespace {
+
+
+class Ell : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Ell<>;
+    using Vec = gko::matrix::Dense<>;
+
+    Ell() : rand_engine(42) {}
+
+    void SetUp()
+    {
+        ASSERT_GT(gko::GpuExecutor::get_num_devices(), 0);
+        ref = gko::ReferenceExecutor::create();
+        gpu = gko::GpuExecutor::create(0, ref);
+    }
+
+    void TearDown()
+    {
+        if (gpu != nullptr) {
+            ASSERT_NO_THROW(gpu->synchronize());
+        }
+    }
+
+    std::unique_ptr<Vec> gen_mtx(int num_rows, int num_cols, int min_nnz_row)
+    {
+        return gko::test::generate_random_matrix<Vec>(
+            ref, num_rows, num_cols,
+            std::uniform_int_distribution<>(min_nnz_row, num_cols),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine);
+    }
+
+    void set_up_apply_data(int stride = 0, int max_nonzeros_per_row = 0)
```

Comment:
are you sure this works fine? I don't see `stride` or `max_nonzeros_per_row` being used anywhere in this function...

Reply:
@gflegar 
Fix it.

I have a problem with the ordering of parameters.
The matrix of nonzeros size is `stride X max_nonzeros_per_row`.
We initial ELL format like `Mtx::create(ref, 0, 0, max_nonzeros_per_row, stride);`
Is it better we change the ordering of stride and max_nonzeros_per_row?

Reply:
Actually, I would have `max_nonzeros_per_row` before the `stride` in the parameter list. The reason for that is that the user is more likely to change the default `max_nonzeros_per_row` than the `stride` parameter:

1.  `Mtx::create(ref, m, n)` creates an `m`-by-`n` matrix, with 0 nonzeros, and stride set to `m`.
2.  `Mtx::create(ref, m, n, nnz)` creates an `m`-by-`n` matrix with `nnz` nonzeros, and stride set to `m`.
3.  `Mtx::create(ref, m, n, nnz, stride)` creates an `m`-by-`n` matrix with `nnz` nonzeros, and stride set to `stride`.

If we were to replace the order of parameters, you lose the possibility to specify option 2. easily, and I feel this is the most useful one, definitely more useful than `Mtx::create(ref, m, n, stride)` (a matrix with 0 nonzeros and stride set to `stride`).